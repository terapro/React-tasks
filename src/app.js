import React from 'react';

import {Header} from './js/components/header/index.js'
import {Summary} from './js/components/summary/index.js'
import {Results} from './js/components/results/index.js'
import {Footer} from './js/components/footer/index.js'

const Body =() => (<div className = "wrapper">
    <Header/>
    <Summary />
    <Results/>

</div>);





const App = () => (
    <React.Fragment>
    <Body/>
    <Footer/>
    </React.Fragment>

);

export {App};