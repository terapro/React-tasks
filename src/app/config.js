const appSettings = {
  searchTypesActivityList: { // Shows lis of search parameters. True/false - Checked/Unchecked. Only 1 can be true
    'title': true,
    'genre': false
  },
  headerSearchTitle: 'Find your movie',
  searchInputPlaceholder: 'Start typing...'



};

export {appSettings};