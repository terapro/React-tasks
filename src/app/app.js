import React, {Component} from 'react';

import {moviesDB} from '../app/data'

import {Body} from '../components/body/body.js'

const searchItemsMatchingFromUserToDBFile = { // User items >> matching >> Items in moviesDB file
  'title': 'title',
  'genre': 'genres',
};
const searchKeysAliases = { // only lower case!
  'sci fi': 'science fiction',
  'scifi': 'science fiction',
  'sci-fi': 'science fiction',
  'sf': 'science fiction',
  'melodrama': 'drama',
  'cartoon': "animation"
};

class App extends Component {

  constructor() {
    super();

    this.state = {
      searchPhrase: '',
      searchType: '',
      sortParameters: null, //Items for 'Sort by' field (e.g. 'release date', 'rating')
      filmNumber: 0,
      result: [],
      searchMode: true,
      filmMode: false,
      filmModeGenre: '',
      filmInfo: {}
    };
    //Search function
    this.startSearchFunc = (obj) => { // Main Search function - callback for the "Search button"
      let phrase = obj.searchPhrase.toLowerCase();
      const type = searchItemsMatchingFromUserToDBFile[obj.searchType];
      let result = [];
      this.setState({
        searchPhrase: phrase,
        searchType: obj.type
      });
      console.log('Searching ' + phrase + ' by ' + type);

      // Search by genres
      if (type === 'genres') {
        phrase = searchKeysAliases[phrase] ? searchKeysAliases[phrase] : phrase; // Synonyms for genres
        for (let i = 0; i < moviesDB.data.length; i++) {
          if (moviesDB.data[i][type].some((genre) => genre.toLowerCase() === phrase))
            result.push(moviesDB.data[i]);
        }
      }
      // Search by title
      if (type === 'title') {
        for (let i = 0; i < moviesDB.data.length; i++) {
          if (moviesDB.data[i][type].toLowerCase() === phrase // Search by the direct coincidence
            || moviesDB.data[i][type].toLowerCase().indexOf(phrase) >= 0) { // Search by keywords in the title
            result.push(moviesDB.data[i]);
          }
        }
      }
      this.state.result = result;
    };
    //Search for film mode by genre
    this.SearchInFilmModeByGenre = (genre) => {
      const cond = {
        searchPhrase: genre,
        searchType: 'genre'
      };
      this.startSearchFunc(cond);
    };
    //Sets the film mode
    this.setFilmMode = (el) => {
      const currentFilm = this.findFilmById(el.target.id);
      const genre = currentFilm['genres'][0];
      this.setState({
                      searchMode: false,
                      filmMode: true,
                      filmModeGenre: genre,
                      filmInfo: currentFilm
                    });
      this.SearchInFilmModeByGenre(genre);
      console.log( this.findFilmById(el.target.id));

    };
    //Sets the search mode
    this.setSearchMode = () => {
      this.setState({
        searchMode: true,
        filmMode: false
      });
      console.log('Entered search mode');
    };
    //Searches film by id
    this.findFilmById =(id) => {
      let res = {x:1};
      for (let i =0; i < moviesDB['data'].length; i++) {
        if (moviesDB['data'][i]['id'] == id) {
          res = moviesDB['data'][i];
          break;
        }
      }
      return res;
    }
  }
  render() {
    return (
      <Body
        startSearch={this.startSearchFunc}
        searchResult={this.state.result}
        searchMode={this.state.searchMode}
        filmMode={this.state.filmMode}
        filmModeGenre={this.state.filmModeGenre}
        filmInfo={this.state.filmInfo}
        setFilmModeCallback={this.setFilmMode}
        setSearchModeCallback={this.setSearchMode}
      />
    )
  }
}

export {App};