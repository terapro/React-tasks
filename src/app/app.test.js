import React from 'react';
import {App} from 'src/app/app.js';
import {shallow} from 'enzyme';

const mockObjectToSearch = {
  searchPhrase: 'Star Wars',
  searchType: 'title'
};
const mockGenre = 'action';

const mockFilm = {
  "id": 333339,
  "title": "Ready Player One",
  "tagline": "A better reality awaits.",
  "vote_average": 8.1,
  "vote_count": 617,
  "release_date": "2018-03-28",
  "poster_path": "https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg",
  "overview": "When the creator of a popular video game system dies, a virtual contest is created to compete for his fortune.",
  "budget": 175000000,
  "revenue": 0,
  "genres": [
    "Adventure",
    "Science Fiction",
    "Action"
  ],
  "runtime": 140
};

describe('<App />', () => {
  describe('Rendering', () => {
    it('should render the App', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });
  });
  describe('Methods', () => {
    let instance;
    beforeAll(() => {
      instance = shallow(<App />).instance();
    });
    afterEach(() => {
      jest.clearAllMocks();
    });
    describe('startSearchFunc', () => {
      it('should store searchType to the state', () => {
        instance.state = {
          searchType: '',
          searchPhrase: '',
          result: []
        };
        instance.startSearchFunc(mockObjectToSearch);
        expect(instance.state.searchType).toEqual(mockObjectToSearch.searchType);
      });
      it('should store searchPhrase to the State', () => {
        instance.state = {
          searchType: '',
          searchPhrase: '',
          result: []
        };
        instance.startSearchFunc(mockObjectToSearch);
        expect(instance.state.searchPhrase).toEqual(mockObjectToSearch.searchPhrase.toLowerCase());
      });
      it('should found a film and store result to the state', () => {
        instance.state = {
          searchType: '',
          searchPhrase: '',
          result: []
        };
        instance.startSearchFunc(mockObjectToSearch);
        expect(instance.state.result.length).not.toBe(0);
      });
    });
    describe('searchInFilmModeByGenre', () => {
      it('should set searchType as "genres" in the state', () => {
        instance.state = {
          searchType: '',
          searchPhrase: ''
        };
        instance.searchInFilmModeByGenre(mockGenre);
        expect(instance.state.searchType).toEqual('genres');
      });
      it('should set proper searchPhrase in the State', () => {
        instance.state = {
          searchType: '',
          searchPhrase: ''
        };
        instance.searchInFilmModeByGenre(mockGenre);
        expect(instance.state.searchPhrase).toEqual(mockGenre);
      });
    });
    describe('setFilmMode', () => {
      it('should set searchMode in the state as false', () => {
        instance.state = {
          searchMode: true,
          filmMode: false,
          filmModeGenre: '',
          filmInfo: {}
        };
        instance.setFilmMode(mockFilm);
        expect(instance.state.searchMode).toEqual(false);
      });
      it('should set filmMode in the state as false', () => {
        instance.state = {
          searchMode: true,
          filmMode: false,
          filmModeGenre: '',
          filmInfo: {}
        };
        instance.setFilmMode(mockFilm);
        expect(instance.state.filmMode).toEqual(true);
      });
      it('should set proper filmModeGenre in the state', () => {
        instance.state = {
          searchMode: true,
          filmMode: false,
          filmModeGenre: '',
          filmInfo: {}
        };
        instance.setFilmMode(mockFilm);
        expect(instance.state.filmModeGenre).toEqual(mockFilm['genres'][0]);
      });
      it('should set proper filmInfo object in the state', () => {
        instance.state = {
          searchMode: true,
          filmMode: false,
          filmModeGenre: '',
          filmInfo: {}
        };
        instance.setFilmMode(mockFilm);
        expect(instance.state.filmInfo).toEqual(mockFilm);
      });
      it('should store the search phrase before toggling from search mode to the film mode', () => {
        instance.state = {
          searchMode: true,
          filmMode: false,
          lastSearchPhrase: '',
          lastSearchType: '',
          searchPhrase: mockObjectToSearch.searchPhrase.toLocaleLowerCase(),
          searchType: mockObjectToSearch.searchType
        };
        instance.setFilmMode(mockFilm);
        expect(instance.state.lastSearchPhrase).toEqual(mockObjectToSearch.searchPhrase.toLowerCase());
      });
      it('should store the genre before toggling from search mode to the film mode', () => {
        instance.state = {
          searchMode: true,
          filmMode: false,
          lastSearchPhrase: '',
          lastSearchType: '',
          searchPhrase: mockObjectToSearch.searchPhrase.toLocaleLowerCase(),
          searchType: mockObjectToSearch.searchType
        };
        instance.setFilmMode(mockFilm);
        expect(instance.state.lastSearchType).toEqual(mockObjectToSearch.searchType);
      });
    });
    describe('setSearchMode', () => {
      it('should set searchMode in the state as true', () => {
        instance.state = {
          searchMode: false,
          filmMode: true
        };
        instance.setSearchMode();
        expect(instance.state.searchMode).toEqual(true);
      });
      it('should set filmMode in the state as false', () => {
        instance.state = {
          searchMode: false,
          filmMode: true
        };
        instance.setSearchMode();
        expect(instance.state.filmMode).toEqual(false);
      });
      it('it should restore the search phrase before toggling from film mode to the search mode', () => {
        instance.state = {
          lastSearchPhrase: mockObjectToSearch.searchPhrase.toLocaleLowerCase(),
          lastSearchType: mockObjectToSearch.searchType,
          searchPhrase: '',
          searchType: ''
        };
        instance.setSearchMode();
        expect(instance.state.searchPhrase).toEqual(mockObjectToSearch.searchPhrase.toLowerCase());
      });
      it('it should restore the genre before toggling from film mode to the search mode', () => {
        instance.state = {
          lastSearchPhrase: mockObjectToSearch.searchPhrase.toLocaleLowerCase(),
          lastSearchType: mockObjectToSearch.searchType,
          searchPhrase: '',
          searchType: ''
        };
        instance.setSearchMode();
        expect(instance.state.searchType).toEqual(mockObjectToSearch.searchType);
      });
    });
  });
});
