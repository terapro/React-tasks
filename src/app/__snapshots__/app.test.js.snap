// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Rendering should render the App 1`] = `
<ErrorBoundary>
  <Body
    filmInfo={Object {}}
    filmMode={false}
    filmModeGenre=" "
    searchInFilmModeByGenre={[Function]}
    searchMode={true}
    searchPhrase=""
    searchResult={Array []}
    setFilmMode={[Function]}
    setSearchModeCallback={[Function]}
    startSearch={[Function]}
  />
</ErrorBoundary>
`;
